GVFS/GVFS.Virtualization/Projection/GitIndexProjection.cs

```diff
@@ -529,12 +532,25 @@ public FileSystemTaskResult AddMissingModifiedFiles()
             {
                 if (this.modifiedFilesInvalid)
                 {
-                    FileSystemTaskResult result = this.indexParser.AddMissingModifiedFiles(this.context.Tracer, this.indexFileStream);
+                    Dictionary<string, PlaceholderListDatabase.PlaceholderData> placeholderDataByPath;
+                    List<PlaceholderListDatabase.PlaceholderData> placeholdersToKeep;
+                    this.placeholderList.GetAllEntries(out placeholderDataByPath, folderPlaceholders: out placeholdersToKeep);
+
+                    FileSystemTaskResult result = this.indexParser.AddMissingModifiedFiles(this.context.Tracer, this.indexFileStream, placeholderDataByPath, placeholdersToKeep);
                     if (result == FileSystemTaskResult.Success)
                     {
                         this.modifiedFilesInvalid = false;
                     }
 
+                    // Any paths that are not in the the index must be added to ModifiedPaths.dat so that git
+                    // reports them as untracked
+                    foreach (string path in placeholderDataByPath.Keys)
+                    {
+                        this.AddModifiedPath(path);
+                    }
+
+                    this.placeholderList.WriteAllEntriesAndFlush(placeholdersToKeep);

```

Comment:
Is this a replace? Or an addition to the existing list? And why do we need to add to the list of placeholders now? I thought we just needed to update the ModifiedPaths?

Reply:
There were two reasons for this:

- The design of `placeholderList` requires that each call to `GetAllEntries` be paired with a corresponding call to `WriteAllEntriesAndFlush`.
- As we add the entries to ModifiedPaths we must remove them from the placeholder list.  If we did not, we could run into a situation like this:

1. Add files to ModifiedPaths.dat here (and leave them in placeholder list)
2. Checkout a branch that does not have the untracked files
3. The usual UpdatePlaceholders logic runs, and it does not find these placeholders in the projection and decides they should be deleted
4. The delete succeeds because the files are still placeholders (i.e. they are not full).

However, all that said, I think we can probably remove the first limitation mentioned above (that each Get requires a Write).  This was originally added so that the collapsing of the placeholder list that occurs during a projection change does not drop and placeholders created during that time.

However, we don't need to collapse the list here (we could just write deletes for the files that should be removed), and so it would be a nice speed up to eliminate the restriction.

Reply:
I don't know if this answered my question, but I think I can infer it - you're saying that we're not _adding_ to the placeholder list here, we're actually removing from it? That was not obvious from the code, but I'm also not totally sold on it if that is true. Still processing...

Reply:
And what does it mean to have placeholder files on disk but not in out PlaceholderList database?

Reply:
> I don't know if this answered my question, but I think I can infer it - you're saying that we're not adding to the placeholder list here, we're actually removing from it? 

Yep that's right, and I think it will be more obvious with the change discussed above (to not call `WriteAllEntriesAndFlush` and to instead call `RemoveAndFlush`).

We remove from the placeholder list by doing the following (and a higher level):

1. Get a list of all placeholders
2. Remove some from that (in memory) list
3. Replace what's on disk with the (updated) list

The "remove" happens by moving items from `placeholderDataByPath` to `placeholdersToKeep` as we decide we want to keep them.  `placeholdersToKeep` starts with all of the folder placeholders as we know from the start we want to keep those.

> And what does it mean to have placeholder files on disk but not in out PlaceholderList database?

It means that we'll still hydrate them (if needed), but we won't update them when the projection changes (because git will).  If we wanted a more detailed name, we could call the list `PlaceholdersOnDiskThatGitIsNotTracking`.  In other words, the database is the intersection of placeholders on disk and index entries that have their 

We've actually supported this state for a long time.  When a placeholder file is renamed:
- It remains a placeholder
- It is added to ModifedPaths.dat
- It is removed from the placeholder list

Reply:
Thanks that all makes sense and I think with your proposed renames, and updating the pattern to remove from, rather than replace, the placeholder list, this logic will become a lot clearer